File: Python/algorithmic_heights/25_ps.py
# Partial Sort
import sys


def convert_to_intlist(line):
    return [int(item) for item in line.split()]


def get_index_of_parent_children_minimum(array, parent_ix, size):
    min_ix = parent_ix
    left_child_ix = 2*parent_ix + 1
    if left_child_ix < size and array[left_child_ix] < array[min_ix]:
        min_ix = left_child_ix

    right_child_ix = left_child_ix + 1
    if right_child_ix < size and array[right_child_ix] < array[min_ix]:
        min_ix = right_child_ix

    return min_ix


def sift_down(array, parent_ix, size):
    while (min_ix := get_index_of_parent_children_minimum(array, parent_ix, size)) != parent_ix:
        array[min_ix], array[parent_ix] = array[parent_ix], array[min_ix]
        parent_ix = min_ix


def heapify(array, size):
    parent_ix = size // 2 - 1
    while parent_ix >= 0:
        sift_down(array, parent_ix, size)
        parent_ix -= 1


def partial_sort(array, size, k):
    heapify(array, size)
    n = size
    while n > size - k:
        array[0], array[n-1] = array[n-1], array[0]
        sift_down(array, 0, n-1)
        n -= 1
    return array[:-(k+1):-1]


def main():
    reader = sys.stdin
    n = int(next(reader))
    array = convert_to_intlist(next(reader))
    k = int(next(reader))
    result = partial_sort(array, n, k)
    print(' '.join(map(str, result)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/25_ps.hs
import Control.Monad (mapM_, when)
import Control.Monad.ST (runST)
import Data.Vector (Vector, freeze, fromList, thaw, toList)
import qualified Data.Vector as V
import qualified Data.Vector.Mutable as MV

convertToIntVector :: String -> Vector Int
convertToIntVector = fromList . map read . words


partialSort :: Ord a => Vector a -> Int -> Int -> Vector a
partialSort vector n k =
    let getIndexOfParentChildrenMinimum array parentIx n = do
            let leftChildIx = 2 * parentIx + 1
            let rightChildIx = leftChildIx + 1
            minIx <- if leftChildIx < n then do
                        x <- MV.read array leftChildIx
                        y <- MV.read array parentIx
                        return $ if x < y then leftChildIx else parentIx
                    else return parentIx
            if rightChildIx < n then do
                x <- MV.read array rightChildIx
                y <- MV.read array minIx
                return $ if x < y then rightChildIx else minIx
            else return minIx

        siftDown array parentIx n =
            let go currentParentIx = do
                    minIx <- getIndexOfParentChildrenMinimum array currentParentIx n
                    when (minIx /= currentParentIx) $ do
                        MV.swap array minIx currentParentIx
                        go minIx
            in go parentIx

        createHeap array parentIx
            | parentIx >= 0 = do
                siftDown array parentIx n
                createHeap array (parentIx - 1)
            | otherwise = return ()

        moveElemToItsPlace array k = do
            MV.swap array 0 (k - 1)
            siftDown array 0 (k - 1)

    in runST $ do
        array <- thaw vector
        createHeap array (n `div` 2 - 1)
        mapM_ (moveElemToItsPlace array) [n, (n-1)..(n-k+1)]
        V.reverse <$> freeze (MV.slice (n - k) k array)


main :: IO ()
main = do
    n <- readLn
    vector <- convertToIntVector <$> getLine
    k <- readLn
    let result = partialSort vector n k
    putStrLn $ unwords $ map show $ toList result
